lib/marisa/grimoire/vector/vector.h

```diff
@@ -231,8 +235,19 @@ class Vector {
     MARISA_DEBUG_IF(new_buf.get() == NULL, MARISA_MEMORY_ERROR);
     T *new_objs = reinterpret_cast<T *>(new_buf.get());
 
-    for (std::size_t i = 0; i < size_; ++i) {
-      new (&new_objs[i]) T(objs_[i]);
+#if __cplusplus >= 201103L
+    constexpr bool trivially_copyable = std::is_trivially_copyable<T>::value;
+#else
+    const bool trivially_copyable = false;
+#endif
+    if (trivially_copyable) {
```

Comment:
If we do this with SFINAE we can avoid the `trivially_copyable` bool and including `<cstring>` for C++98.

```c++
#if __cplusplus >= 201103L
template <typename T, std::enable_if<std::is_trivially_copyable<T>::value, int> = 0>
void realloc_move(T *dest, const T *src, std::size_t size) {
  std::memcpy(reinterpret_cast<void*>(dest),
              reinterpret_cast<const void*>(src),
              sizeof(T) * src);
}

template <typename T, std::enable_if<!std::is_trivially_copyable<T>::value, int> = 0>
#endif
void realloc_move(T *dest, const T *srcs, std::size_t size) {
  for (std::size_t i = 0; i < size; ++i) {
    new (&new_objs[i]) T(objs[i]);
  }
  for (std::size_t i = 0; i < size; ++i) {
    objs[i].~T();
  }
}
```

Reply:
This greatly increases the level of C++ knowledge required to understand the code. All to avoid one `#include`?

Reply:
Fair enough. I find the template easier to read but it's subjective.

As this code already uses placement-new and calls destructors explicitly I think we can assume that the reader is an expert C++ user :)

Reply:
More concretely: I don't understand std::enable_if well enough to want to use it in code I write, and I wrote this. :-)